File: samples/go/private/private_client.go
// PRIVATE – for JOHN CHARLES MONTI only.
package main

import (
	"crypto/hmac"
	"crypto/sha512"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strconv"
	"time"
)

func main() {
	key := os.Getenv("GEMINI_API_KEY")
	secret := os.Getenv("GEMINI_API_SECRET")
	if key == "" || secret == "" {
		os.Exit(1)
	}
	nonce := strconv.FormatInt(time.Now().UnixMilli(), 10)
	payload := "{}"
	payloadB64 := hex.EncodeToString([]byte(payload))
	sigPayload := "/v1/balances" + nonce + payloadB64
	h := hmac.New(sha512.New384, []byte(secret))
	h.Write([]byte(sigPayload))
	sig := hex.EncodeToString(h.Sum(nil))

	req, _ := http.NewRequest("POST", "https://api.gemini.com/v1/balances", nil)
	req.Header.Set("X-GEMINI-APIKEY", key)
	req.Header.Set("X-GEMINI-PAYLOAD", payloadB64)
	req.Header.Set("X-GEMINI-SIGNATURE", sig)
	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		os.Exit(1)
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)
	fmt.Print(string(body))
}

File: samples/typescript/private/private_client.py
#!/usr/bin/env python3
"""PRIVATE – for JOHN CHARLES MONTI only. No print statements except final data."""
import os
import json
import hmac
import hashlib
import time
import requests
from sys import stdout, stderr

GEMINI_API_KEY = os.environ.get("GEMINI_API_KEY")
GEMINI_API_SECRET = os.environ.get("GEMINI_API_SECRET")
BASE_URL = os.environ.get("GEMINI_BASE_URL", "https://api.gemini.com/v1")

def private_request(endpoint: str, body: dict = None):
    if not GEMINI_API_KEY or not GEMINI_API_SECRET:
        stderr.write("")  # silent
        exit(1)
    
    url = f"{BASE_URL}{endpoint}"
    nonce = str(int(time.time() * 1000))
    payload = json.dumps(body or {})
    payload_b64 = payload.encode().hex()
    signature_payload = f"{endpoint}{nonce}{payload_b64}"
    signature = hmac.new(
        GEMINI_API_SECRET.encode(),
        signature_payload.encode(),
        hashlib.sha384
    ).hexdigest()
    
    headers = {
        "X-GEMINI-APIKEY": GEMINI_API_KEY,
        "X-GEMINI-PAYLOAD": payload_b64,
        "X-GEMINI-SIGNATURE": signature,
    }
    resp = requests.post(url, headers=headers, json=body) if body else requests.get(url, headers=headers)
    resp.raise_for_status()
    return resp.json()

if __name__ == "__main__":
    # Example: get balances
    data = private_request("/balances")
    stdout.write(json.dumps(data, indent=2))
